### PR TITLE
New performance recipes document

### LINKED ISSUE #10606: Create performance troubleshooting recipes document
In addition to the existing tutorials we use that are more focused on schema design:
https://www.cockroachlabs.com/docs/v21.1/make-queries-fast.html
https://www.cockroachlabs.com/docs/v21.1/sql-tuning-with-explain.html

we want a recipes document that would have small, targeted performance troubleshooting scenarios with actions the user can perform. For example, the user notices a high level of contention in the SQL metrics table in the DB Console. What should they do?

The recommendations must be high-level, as each application and deployment is going to have its own particular steps. The idea is to point the user in the right direction.

Brainstorming document: https://docs.google.com/document/d/1IwYbqdaQ9BYAxWhnQ1KOdYkZ3IX91-dbI7f9NQA5z94/edit?usp=sharing

cc @awoods187 @kevin-v-ngo

### PR DESCRIPTION
**Pull request by ianjevans**
Fixes #10606.

New document to direct users to possible solutions to common performance problems.

### CODE REVIEW

File: v21.1/performance-recipes-solutions.md
Comment (kevin-v-ngo) [inline v21.1/performance-recipes-solutions.md:29]: This should be the Transaction Restarts chart which is now on the DB console SQL metrics page (used to be on the Distributed metrics dashboard). 

https://github.com/cockroachdb/cockroach/commit/ef5135df0e514552eee99bad7264132508cf3f41

File: v21.1/performance-recipes-solutions.md
Comment (kevin-v-ngo) [inline v21.1/performance-recipes-solutions.md:77]: Same comment as below. Perhaps we should call out per node? This scenario would lead to hot spots/nodes.

File: v21.1/performance-recipes-solutions.md
Comment (kevin-v-ngo) [inline v21.1/performance-recipes-solutions.md:61]: There's another signal/problem to include:

The Full Table/Index Scans chart in the DB console SQL metrics dashboard is showing spikes over time.

**Review by kevin-v-ngo — CHANGES_REQUESTED** [review]
Looks good and added comments. Let's sync on next steps and what you need from me.

**Comment by jseldess** [comment]
@ianjevans, any updates on this PR?

**Comment by ianjevans** [comment]
@jseldess I've been waiting on @kevin-v-ngo for a final review, and had brought it up in my last 1:1 with him. I'll have a writer review it in the meantime, and if there are changes we can update the live docs.

**Comment by kevin-v-ngo** [comment]
Apologies for the delay. It looks good on my end @ianjevans. Note the links in the 'Possible solution' column are not working for me when browsing the preview - not sure if this is expected.

**Review by ericharmeling — CHANGES_REQUESTED** [review]
Nice work. This doc should be super helpful.

I have several comments and suggestions related to style, grammar, etc. All comments apply to both 21.1 and 21.2 versions.

> Reviewed 1 of 5 files at r1, 1 of 1 files at r4, 7 of 7 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10948#-:-Mkbx60nChXmjl2hl0d3:bjjoyph)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ianjevans and @kevin-v-ngo)

___
*[_includes/sidebar-data-v21.1.json, line 415 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbprWl0ed-UFbmajpz:-MkbprWmD5DmYmJp9Rj4:b-nnew8e) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/_includes/sidebar-data-v21.1.json#L415)):*
> ```JSON
>               {
>                 "title": "Use the EXPLAIN statement",
>                 "urls": [
> ```

Should the 21.2 sidebar also be updated?
___
*[_includes/v21.1/performance/statement-contention.md, line 3 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-Mkbp81MDqN0ejIgCZu4:-Mkbp81N0AKwq1RT1YH0:bwtiqa4) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/_includes/v21.1/performance/statement-contention.md#L3)):*
> ```
> DB Console
> ```

Did we decide on the naming for the DB Console? Or are we planning to use variables for this product name?
___
*[_includes/v21.1/performance/statement-contention.md, line 7 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbpJ4O00osvc_cIdpG:-MkbpJ4O00osvc_cIdpH:b-px7j5v) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/_includes/v21.1/performance/statement-contention.md#L7)):*
> ```
> {%comment%} Replace with final statements {%endcomment%}
> ```

Was this done? If so, could remove this comment.
___
*[v21.1/performance-recipes.md, line 21 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbsafMAnVSmQa3udoJ:-MkbsafN5mnht-WLwcuJ:b-z5wzqj) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes.md#L21)):*
> ```
> Your application is experiencing degraded performance with the following serialization errors:<br>`SQLSTATE: 40001`<br>`RETRY_WRITE_TOO_OLD`<br>`RETRY_SERIALIZABLE`<br><br>The [SQL Statement Contention dashboard](ui-sql-dashboard.html#sql-statement-contention) in the DB Console is showing spikes over time.<br><br>The [SQL Statement Errors graph](ui-sql-dashboard.html#sql-statement-errors) in the DB Console is showing spikes in retries over time.
> ```

Should these all be items in an unordered list? If not, there should probably be an `<hr>` to better separate the problems at least.

The same goes for all of the other "Problem" rows.
___
*[v21.1/performance-recipes.md, line 21 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-Mkbt-c-E2IZubvrcpay:-Mkbt-c-E2IZubvrcpaz:b-tssx8j) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes.md#L21)):*
> ```
>  serialization errors
> ```

```
[serialization errors](transaction-retry-error-reference.html)
```
___
*[v21.1/performance-recipes.md, line 22 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbvamIEAnHJP39fD_q:-MkbvamIEAnHJP39fD_r:b-ykws7m) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes.md#L22)):*
> ```
> `EXPLAIN` or `EXPLAIN ANALYZE`
> ```

Probably should link to the EXPLAIN docs.
___
*[v21.1/performance-recipes.md, line 22 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-Mkbvf8pE1iC43AyFDbr:-Mkbvf8pE1iC43AyFDbs:btyzqn3) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes.md#L22)):*
> ```
> `CRDB_INTERNAL.node_statement_statistics`
> ```

`crdb_internal.node_statement_statistics`
___
*[v21.1/performance-recipes.md, line 22 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbvlMv2HNg2m-LUChO:-MkbvlMv2HNg2m-LUChP:b-fuz0w0) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes.md#L22)):*
> ```
> `SHOW FULL TABLE SCANS`
> ```

```
[`SHOW FULL TABLE SCANS`](show-full-table-scans.html)
```
___
*[v21.1/performance-recipes-solutions.md, line 27 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbqnXD2FGuJ8rAnXD6:-MkbqnXD2FGuJ8rAnXD7:bt2c285) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L27)):*
> ```
> {%comment%} Add screenshot of high-contention graph {%endcomment%}
> ```

Is this still planned?
___
*[v21.1/performance-recipes-solutions.md, line 29 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbqaII3E5kfORuwMX3:-MkbqaII3E5kfORuwMX4:b-47bvfm) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L29)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> Your application is experiencing degraded performance with serialization errors like:
> 
>   - `SQLSTATE: 40001`
>   - `RETRY_WRITE_TOO_OLD`
>   - `RETRY_SERIALIZABLE`
> 
> The [SQL Statement Contention graph](ui-sql-dashboard.html#sql-statement-contention) graph is showing spikes over time.
> 
> {%comment%} Add screenshot of high-contention graph {%endcomment%}
> 
> The [KV Transaction Restarts graph](ui-overview.html) graph is showing spikes in retries over time.
> ```
</details>


Should these all be in an unordered list?
___
*[v21.1/performance-recipes-solutions.md, line 31 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbqieK34atR_SKRe0c:-MkbqieK34atR_SKRe0d:b98v9ss) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L31)):*
> ```
> {%comment%} Add screenshot of high-retries graph {%endcomment%}
> ```

Is this still planned?
___
*[v21.1/performance-recipes-solutions.md, line 35 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbrUmD6jEuVKScfmhi:-MkbrUmD6jEuVKScfmhj:b-uzb1xg) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L35)):*
> ```
> {% include {{ page.version.version }}/performance/statement-contention.md %}
> ```

The instructions in this file should probably be ordered steps, with a callout note or two. Right now, it's just sentences with line breaks, which is hard to follow when there are multiple instructions.
___
*[v21.1/performance-recipes-solutions.md, line 41 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-Mkbxud43uTvPO0UMCt9:-Mkbxud43uTvPO0UMCtA:b-8i8y3t) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L41)):*
> ```Markdown
> 
> ## Indicators that your workload has statements with full table scans
> ```

Same comment re: unordered list here.
___
*[v21.1/performance-recipes-solutions.md, line 60 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbrnZr0sxXkTjMzn6_:-MkbrnZr0sxXkTjMzn6a:b-uzi9g4) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L60)):*
> ```
> If the statement plans returned by the [`EXPLAIN`](sql-tuning-with-explain.html) and [`EXPLAIN ANALYZE` commands](explain-analyze.html) indicate that there are full table scans.
> ```

This is a sentence fragment (it's just a conditional clause). The sentence structure should match that of the other items in this list of indicators (i.e., complete sentences).
___
*[v21.1/performance-recipes-solutions.md, line 72 at r5](https://reviewable.io/reviews/cockroachdb/docs/10948#-MkbwbnE2uwfgh2Z1srC:-MkbwbnE2uwfgh2Z1srD:b6usvxw) ([raw file](https://github.com/cockroachdb/docs/blob/6ae11eed607c644de1a1ed2f564f3853c1c2c516/v21.1/performance-recipes-solutions.md#L72)):*
> ```Markdown
> <section class="filter-content" markdown="1" data-scope="primarykey">
> 
> ## Indicators that your tables are using suboptimal primary keys
> ```

Same suggestions here as above:

- The indicators should be part of an unordered list (i.e., a bulleted list of separate, possible indicators).
- The solutions should be part of an ordered list (i.e., steps to solve the problem).

**Comment by ianjevans** [comment]
> Should the 21.2 sidebar also be updated?

Done.

> Did we decide on the naming for the DB Console? Or are we planning to use variables for this product name?

I don't believe we have a variable yet for this, though we should eventually. I'm pretty sure we landed on "CockroachDB Console" and CockroachDB Cloud Console" as official names, but I don't know if we have the OK to use "DB Console" and "Cloud Console" as short names. When that happens, we should do a thorough pass throughout the entire docset.

> Was this done? If so, could remove this comment.

The v21.2 queries have been updated, but I will confirm with Kevin about the v21.1 queries. Also removed the comments about screenshots as I don't know if we need them.

> Should these all be items in an unordered list? If not, there should probably be an <hr> to better separate the problems at least. The same goes for all of the other "Problem" rows.

Complex table support in Markdown is terrible, and I was trying to avoid having to reformat everything in raw HTML, but I guess I can't avoid it. Done.

> [various linking comments]

Done.

> The instructions in this file should probably be ordered steps, with a callout note or two. Right now, it's just sentences with line breaks, which is hard to follow when there are multiple instructions.

They're not ordered steps, they're hints with links off to the actual instructions. I've made that more clear now. It makes more sense when you're seeing the rendered HTML. The main purpose of this doc is to narrow down and give users more direction on how to find the more detailed solutions to their complicated performance problems.

> This is a sentence fragment (it's just a conditional clause). The sentence structure should match that of the other items in this list of indicators (i.e., complete sentences).

Done.

**Review by ericharmeling — APPROVED** [review]
Looks great! 

The tables thing is a real problem - thanks for doing that in HTML. Sorry for making things more complicated here... :robot: 

> Reviewed 6 of 7 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10948#-:-MkcjYpx9Bu88dgYlQZo:b-xbwe8o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ianjevans and @kevin-v-ngo)

**Pull request merged**
